src/main/scala/io/iohk/ethereum/ledger/InMemoryWorldStateProxy.scala

```diff
@@ -172,6 +175,23 @@ class InMemoryWorldStateProxy private(
   override def saveStorage(address: Address, storage: InMemoryWorldStateProxyStorage): InMemoryWorldStateProxy =
     copyWith(contractStorages = contractStorages + (address -> storage.wrapped))
 
+  override def touchAccounts(addresses: Address*): InMemoryWorldStateProxy =
+    copyWith(touchedAccounts = touchedAccounts.map(oldAddresses => oldAddresses ++ addresses.toSet))
```

Comment:
what if `touchedAccounts` is None? is it expected that it remains `None` in this case?

Reply:
oh I see now, `touchedAccounts` also holds the boolean flag `noEmptyAccounts`. What do you think about making it more explicit and adding this flag separately? (and changing `touchedAccounts` to just `Set`)

Reply:
`touchedAccounts` is set based on `noEmptyAccounts` flag in https://github.com/input-output-hk/mantis/pull/306/files#diff-7632a397485d58b59d376a5eb28a40c8R36. How do you think it could be made more explicit ?

And  Actually, `touchedAccounts`  could be just `Set` always, and we could always just add accounts to it, delete empty accounts only if a flag is set, and clear it at the end of the transaction. But i thought that it would be quite a wasteful approach as: 
1. `oldAddresses ++ addresses.toSet` operations on `Set` are quite expensive. 
2. This eip is only enabled for ETH net, so we would be wasting resources thorugh all ETC chain. 

My logic here was that making it `Option[Set]`, we : 
explicitly declare that sometimes (on ETC chain, before fork block), we don't need those touched accounts.
Save ourselves quite a computation that is not necessary, and `Option` take care of all `if .. else` comparisons.

Reply:
Ok, let's leave it as an Option

Reply:
Well, I don't know... I like @LukasGasior1 's idea of separating those aspects. And while it's a valid point that want to avoid unnecessary `setA ++ setB` - why would we have to do that?

```
override def touchAccounts(addresses: Address*): InMemoryWorldStateProxy =
  if (noEmptyAccounts)
    copyWith(touchedAccounts = touchedAccounts ++ addresses.toSet)
  else 
    this
```

That should simplify the related code, and I don't think you'd need another conditional on `noEmptyAccounts` flag.

Reply:
You have valid point with that there would be only conditional needed, but I am still not entirely convinced by that approach as: 
1. We have other methods like `combineTouchedAccounts`,   `clearTouchedAccounts`, we would not need to have conditionals in them, but they would operate on empty sets mostly. ( not a deal breaker) 
2.  In `InMemoryWorldStateProxy` we would have something like 
`  val touchedAccounts: Set[Address],
  val noEmptyAccount: Boolean`
in a constructor and it is not immediately obvious that these values are correlated with each other.
Only look at `touchAccounts` would reveal that these values exist together.
now this dependence is explicitly declared in `apply` method

Reply:
1. Yes, and btw `EmptySet` is optimised
2. I agree that without any documentation it's not obvious, just as it's not obvious why `Option[Set[..]]` is used. Code comments should clarify that

Reply:
1. This is not a matter of optimisation, but a matter of some silent assumption make by other functions that they are operating on empty sets, `Option` takes this logic out in open
2. Good point with adding some comments. Nevertheless `if (noEmptyAccounts) Some(Set.empty) else None` constructor makes it obvious when need or dont need the `Set[..]`. ( though i can be biased here as i am implenter and you and @LukasGasior1  did not see it that way )

Reply:
The more I think about it, the more it feels as the matter of preference of representation 
(`Option[Set] vs  Set and Boolean`). 
So let's have it your way as there is 2 vs 1, in terms who prefers what

Reply:
@KonradStaniec don't get me wrong. There was nothing fundamentally wrong with the `Option` solution, and the proposed change is only cosmetic. But I disagree about it being a matter of preference. Separate set and flag are slightly, but clearly better because:

- philosophical argument: understanding 2 simple things is easier than 1 thing composed of 2 things. Terms _separation of concerns_ applies to variables as well.
- practical and more important argument: in most places you don't need to deal with both aspects - you're only interested in the elements of the set, and not whether they are collected at all - which results in cleaner simpler code.